pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java

```diff
@@ -119,10 +119,12 @@ public ConsumerBuilderImpl(PulsarClientImpl client, Schema<T> schema) {
             return FutureUtil.failedFuture(
                     new InvalidConfigurationException("KeySharedPolicy must set with KeyShared subscription"));
         }
-        if(conf.isRetryEnable() && conf.getTopicNames().size() > 0 ) {
-            TopicName topicFirst = TopicName.get(conf.getTopicNames().iterator().next());
-            String retryLetterTopic = topicFirst.getNamespace() + "/" + conf.getSubscriptionName() + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX;
-            String deadLetterTopic = topicFirst.getNamespace() + "/" + conf.getSubscriptionName() + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX;
+        if (conf.isRetryEnable() && conf.getTopicNames().size() > 0 ) {
+            String topicName = TopicName.get(conf.getTopicNames().iterator().next()).getPartitionedTopicName();
+            String retryLetterTopic = topicName + "-" + conf.getSubscriptionName()
+                    + RetryMessageUtil.RETRY_GROUP_TOPIC_SUFFIX;
+            String deadLetterTopic = topicName + "-" + conf.getSubscriptionName()
+                    + RetryMessageUtil.DLQ_GROUP_TOPIC_SUFFIX;
```

Comment:
Topics might not belong to a partitioned topic since we can use `.topic("persistent://my-tenant/my-ns/topic-a", "persistent://my-tenant/my-ns/topic-b", "persistent://my-tenant/my-ns/topic-c")`. This is why the default retry topic name does not contains the topic(the topic name is written to the properties of the message when send it to the retry topic).

You can explicitly specify the retry topic name or the dead letter topic name through the dead letter policy.

Reply:
Thanks for your explain. I got that why the original default retry letter topic does not contain  the topic.  There is no doubt that the best way to avoid this problem is specify the retry letter topic and dead letter topic. 

In out prduction environment, we maintain a public pulsar cluster and many users access it. There are some users just set enableRetry to true without specify retry letter topic and dead letter topic which may lead to problem discribed by #9742.

IMO, it's better we can find a way to fix this problem. Maybe we can handle default retry and dead letter topic for each topic.

> Topics might not belong to a partitioned topic since we can use .topic("persistent://my-tenant/my-ns/topic-a", "persistent://my-tenant/my-ns/topic-b", "persistent://my-tenant/my-ns/topic-c").
when consumer subscribe multi topics, pulsar can set the default retry letter topic and dead letter topic for every topic seperatlly, i.e. for `persistent://my-tenant/my-ns/topic-a` will have default retry letter topic `persistent://my-tenant/my-ns/topic-a-subscription-name-RETRY` and dead letter topic `persistent://my-tenant/my-ns/topic-a-subscription-name-DLQ`;  for `persistent://my-tenant/my-ns/topic-b` will have default retry letter topic `persistent://my-tenant/my-ns/topic-b-subscription-name-RETRY` and dead letter topic `persistent://my-tenant/my-ns/topic-b-subscription-name-DLQ`

Reply:
Make sense to me. And I think after this change will break the users that uses the try letter topic due to the different topic name. So this should discuss it in the dev email channel and highlight this will be a breaking change, so that we can highlight it at the release note.